qiime2/core/archive/format/v5.py

```diff
@@ -25,8 +29,39 @@ def write(cls, archive_record, type, format,
 
     @classmethod
     def write_checksums(cls, archive_record):
-        write_checksums(
-            directory=str(archive_record.root),
-            checksum_file=cls.CHECKSUM_FILE,
-            checksum_type=cls.CHECKSUM_TYPE
-        )
+        checksum_cache = ChecksumCache()
+
+        checksums = {}
+        for root, dirs, files in os.walk(archive_record.root, topdown=True):
+            dirs[:] = sorted([d for d in dirs if not d[0] == '.'])
+            for file in sorted(files):
+                if file[0] == '.':
+                    continue
+
+                path = Path(root) / file
+                archive_relative_path = path.relative_to(
+                    Path(archive_record.root)
+                )
+
+                # check checksum cache
+                checksum_string = None
+                if archive_relative_path.is_relative_to(
+                    Path("provenance/artifacts")
+                ):
+                    cached_path = archive_relative_path.relative_to(
+                        Path("provenance/artifacts")
+                    )
+                    checksum_string = checksum_cache.get(cached_path)
+
+                if checksum_string is None:
+                    checksum_string = checksum(str(path), cls.CHECKSUM_TYPE)
+
+                checksums[str(archive_relative_path)] = checksum_string
+
+        with (Path(archive_record.root) / cls.CHECKSUM_FILE).open('w') as fh:
+            for item in checksums.items():
+                # always ignore annotations dir when writing checksums
+                # so we dont self-invalidate when adding/removing annotations
+                if not item[0].startswith('annotations'):
```

Comment:
Should we skip even calculating the checksums for annotations? Right now it looks like we calculate the checksums for the annotations then never write them or do anything with them, because as the comment says annotations are mutable.